src/Xrefcheck/Verify.hs

```diff
@@ -231,22 +274,39 @@ data NeedsCaching key
 -- then the @action@ will not be executed, and the value is added to the accumulator list.
 -- After the whole list has been traversed, the accumulator is traversed once again to ensure
 -- every asynchronous action is completed.
+-- If interrupted by AsyncException, returns this exception and list of already calcualted results
+-- (their subset can be arbitrary). Computations that were not ended till this moment are cancelled.
 forConcurrentlyCaching
   :: Ord cacheKey
-  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO [b]
-forConcurrentlyCaching list needsCaching action = go [] M.empty list
+  => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO (Either (AsyncException, [b]) [b])
+forConcurrentlyCaching list needsCaching action = do
+  res <- go [] M.empty list
+  case res of
+    Right r -> return $ Right $ map (fromMaybe (error "impossible")) r
```

Comment:
I think we can avoid this `error "impossible"` case if we use [`poll`](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html#v:poll) (which returns a `(Maybe (Either SomeException a))`), instead of `wait` (which returns `a`).

Consider:

```diff
 forConcurrentlyCaching
   :: forall a b cacheKey. Ord cacheKey
   => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> IO (Either (AsyncException, [b]) [b])
 forConcurrentlyCaching list needsCaching action = do
-  res <- go [] M.empty list
-  case res of
-    Right r -> return $ Right $ map (fromMaybe (error "impossible")) r
-      -- impossible, since no actions were cancelled
-
-    Left (exception, partialResult) -> return $  Left (exception, catMaybes partialResult)
+  go [] M.empty list
   where
-    interruptibleAction :: a -> IO (Maybe b)
-    interruptibleAction a = handle (\AsyncCancelled -> pure Nothing) (Just <$> action a)
-
     go
-      :: [Async (Maybe b)]
-      -> Map cacheKey (Async (Maybe b))
+      :: [Async b]
+      -> Map cacheKey (Async b)
       -> [a]
-      -> IO (Either (AsyncException, [Maybe b]) [Maybe b])
+      -> IO (Either (AsyncException, [b]) [b])
     go acc cached items = handle
       (\exception -> do
           partialResults <- for acc \asyncAction -> do
             cancel asyncAction
-            wait asyncAction
-          pure $ Left (exception, partialResults)
+            poll asyncAction <&> \case
+              Nothing -> Nothing
+              Just (Right a) -> Just a
+              Just (Left _ex) -> Nothing
+          pure $ Left (exception, catMaybes partialResults)
       )
       -- If action was already completed, then @cancel@ will have no effect, and we
       -- will get a result from @cancel f >> wait f@. Otherwise interrupted action
       -- will return us @Nothing@.
       case items of
         [] -> Right . reverse <$> for acc wait
 
         (x : xs) -> case needsCaching x of
           NoCaching -> do
-            withAsync (interruptibleAction x) $ \b ->
+            withAsync (action x) $ \b ->
               go (b : acc) cached xs
           CacheUnderKey cacheKey -> do
             case M.lookup cacheKey cached of
               Nothing -> do
-                withAsync (interruptibleAction x) $ \b ->
+                withAsync (action x) $ \b ->
                   go (b : acc) (M.insert cacheKey b cached) xs
               Just b -> go (b : acc) cached xs

```